### PR TITLE
Packaging and release blog post

### PR DESCRIPTION
**Pull request by mehrdada**

### CODE REVIEW

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+gRPC is [developed in the open on GitHub](https://github.com/grpc)
+and the `master` branch undergoes continuous testing to try to ensure the tip of `master` is buildable and passes the tests.  In this post, we discuss the release process of gRPC as an open source project to clarify it for our users,
+and to serve as an example for other open source projects to consider adopting.  This article is written mostly from the perspective of the gRPC Core (C-based) languages.  The release processes of gRPC Java and Go, although similar in spirit, may be slightly different in technicalities.
+
+## Official Releases ##
+
+gRPC follows semantic versioning and we aim to ship a new minor version every six weeks.  While `master` is quite good and heavily tested, and we could almost always simply consider tagging its `HEAD` as a minor release whenever we like to ship one, a more rigorous quality control process is justified for 
+supported gRPC releases, as it is a critical piece of infrastructure for many users.  As a consequence, we schedule a feature freeze every six weeks.
+[gRPC Release Schedule]() highlights the actual dates.
```
Comment (carl-mastrangelo) [inline _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: Should this be a link?

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+systems like [Bazel](https://bazel.build) improve the build from source
+and distribution experience for scenarios where it is supported, setting
+up the infrastructure to build gRPC from source can be a hurdle for
+some people who like to experiement with the bleeding edge gRPC code
+on the `master` branch.  To alleviate this concern for the users who
+want to be on the bleeding edge, or just cannot wait for a particular
+bug fix or feature on `master` to land in an official release, as well
+as facilitating bisection while debugging by providing prebuilt binaries
+at regular intervals, [we produce packages](https://github.com/grpc/grpc/tree/master/tools/internal_ci/linux/grpc_publish_packages.sh) every day and upload them to
+a [Google Cloud Storage bucket](gs://packages.grpc.io), also
```
Comment (carl-mastrangelo) [inline _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: bad link?

Comment (mehrdada) [reply _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: It's not a bad link. It's a Google Cloud Storage bucket URI. Not loadable in the browser.

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md (hunk 1 above)
Comment (mehrdada) [reply _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: Yes, it should—but @hsaliak should create the destination first.

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+is promptly bumped to reflect that new features will land starting in the followup release.  The `master` branch for gRPC Core repository always carries a `-dev` suffix to indicate it is under active development.
+
+### Prereleases and Release Candidates ###
+
+Once the release branch is cut, its version is updated to reflect its prerelease status, *e.g.* on the Core repository, the suffix will be
+changed to `-pre1` indicating it is the first release candidate of that version (subsequent RCs, if any, will get `-pre2`, *etc.* suffixes.)
+
+A package build job is then triggered on the prerelease branch to create binary packages for the supported languages and platforms.  The binary packages undergo *distribution tests*, which are tests that verify correct installation and loading of the binary packages on supported platforms.  If everything looks fine as expected, the release manager then tags `HEAD` of the release branch as a "prerelease" on GitHub and the prerelease package builds will be uploaded to the common package managers of the supported languages.
+
```
Comment (srini100) [inline _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: How do users sign up to get notifications of RC releases?

**Comment by mehrdada** [comment]
@tdbhacks Any updates on release schedule document that is supposed to be linked here? If not, I'll just remove the link and publish the post.

**Comment by hsaliak** [comment]
@mehrdada please give me a couple of days to send you a PR for that :)

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+company-link: https://www.google.com
+---
+
+# gRPC Packaging: Release Process and Nightly Builds #
+
+gRPC is [developed in the open on GitHub](https://github.com/grpc) and the
+`master` branch undergoes continuous testing to try to ensure the `HEAD` of
+`master` is buildable and passes the tests.  In this post, we discuss the
+release process of gRPC as an open source project to clarify it for our users,
+and to serve as an example for other open source projects to consider adopting.
```
Comment (hsaliak) [inline _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: I would avoid stating that we want this to serve as an example for other open source projects. 
"In this post we discuss the release process of gRPC" should be sufficient..

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+spirit, may be slightly different in technicalities.
+
+## Official Releases ##
+
+gRPC follows semantic versioning and we aim to ship a new minor version every
+six weeks.  While `master` is quite good and heavily tested, and we could
+almost always simply consider tagging its `HEAD` as a minor release whenever we
+like to ship one, a more rigorous quality control process is justified for
+supported gRPC releases, as it is a critical piece of infrastructure for many
+users.  As a consequence, we schedule a feature freeze every six weeks.  [gRPC
```
Comment (hsaliak) [inline _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: for many users and companies..

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+Release Schedule]() highlights the actual dates.
```
Comment (hsaliak) [inline _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: What would be the best place to host this doc, should we have it up somewhere on grpc.io outside of the blog post?

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+The feature window for each minor release closes when a release branch is cut
+from `master`.  The release manager of that particular release takes a commit
+close to the `HEAD`, often the `HEAD` itself, where the tests are green, and
+creates a new release branch at that point.  Starting at the branch cut point,
+additional feature development continues on the `master` branch without
+affecting the release branch and changes to the release branch are kept
+minimal, *i.e.* mostly bug fixes.  The version number on the `master` branch is
+promptly bumped to reflect that new features will land starting in the followup
+release.  The `master` branch for gRPC Core repository always carries a `-dev`
+suffix to indicate it is under active development.
```
Comment (hsaliak) [inline _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: Do you want to state here that the branch cut date is a hard rolling deadline of every 6 weeks?

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+A package build job is then triggered on the prerelease branch to create binary
+packages for the supported languages and platforms.  The binary packages
+undergo *distribution tests*, which are tests that verify correct installation
+and loading of the binary packages on supported platforms.  If everything looks
+fine as expected, the release manager then tags `HEAD` of the release branch as
+a "prerelease" on GitHub and the prerelease package builds will be uploaded to
+the common package managers of the supported languages.
+
+Mission-critical users depending on the latest version of gRPC in their
```
Comment (hsaliak) [inline _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: We strongly encourage users of gRPC for mission-critical systems to..

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+application should run their continuous tests against the latest prerelease, as
+well as the latest release, to see possible regressions early and file issues
+and/or temporarily pin the dependency version to guard against breakages.
+
+After the branch is cut and the prerelease is evaluated in the wild for a week
+or two, the release candidate gets promoted to the final version.  Bug fixes
```
Comment (hsaliak) [inline _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: for at least 2 weeks..?

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+landing in the release branch after the first prerelease induce subsequent
+prerelease and qualification processes (albeit often with a shorter timescale)
+to decrease the risk of regressions the moment a new stable package is uploaded
+to the language package managers.  Strictly adhering to this ritual is critical
```
Comment (hsaliak) [inline _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: s/ritual/process

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md (hunk 6 above)
Comment (mehrdada) [reply _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: I think we can host it as a view-only Google Spreadsheet linked here for now.

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md (hunk 4 above)
Comment (mehrdada) [reply _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: Why?

File: _posts/2018-08-31-grpc-packaging-releases-nightlies.md (hunk 6 above)
Comment (srini100) [reply _posts/2018-08-31-grpc-packaging-releases-nightlies.md]: Point to https://github.com/grpc/grpc/blob/master/doc/grpc_release_schedule.md

**Comment by mehrdada** [comment]
@hsaliak Can you take over this and drive to publication please?

**Comment by hsaliak** [comment]
Yes - i can do that..let me try to get this published this week.

**Comment by thisisnotapril** [comment]
@hsaliak should we keep this one open or go ahead and close it for now? trying to clear out open PRs before website migration.

**Comment by hsaliak** [comment]
We can potentially close this and re-open and drive to completion as suggested at a later date.

**Comment by thisisnotapril** [comment]
> We can potentially close this and re-open and drive to completion as suggested at a later date.

sounds good; going with that!

**Pull request closed without merging**